### PR TITLE
Added demo project: app-test-demo

### PR DESCRIPTION
**Pull request by vilmantas-r**

### CODE REVIEW

File: app-test-demo/src/test/java/com/nortal/test/demo/configuration/TestDemoConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package com.nortal.test.demo.configuration;
+
+import org.springframework.cloud.openfeign.EnableFeignClients;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@EnableFeignClients(basePackages = {"com.nortal.test.demo.api"})
```
Comment (DonMelo) [inline app-test-demo/src/test/java/com/nortal/test/demo/configuration/TestDemoConfiguration.java]: currently feign basePackage is "com.nortal.test.feign.demo.api"

**Review by DonMelo — APPROVED** [review]

File: app-test-demo/README.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
+# Todo Application Test Demo
+
+This project is intended to demonstrate use of multiple Test Automation Framework modules together.
+
+Demo project executes tests against Todo application that consists of frontend, backend and Postgres database.
+
+## Test Automation Framework modules
+* Core - main functionality that ties all other modules.
+* Containers - [TestContainers](https://www.testcontainers.org/) based module that provides a capability to test against a docker container.
+* Jdbc - JDBC support to execute database queries. SQL queries are specified in: `src/test/resources/query`.
```
Comment (ricardas-buc) [inline app-test-demo/README.md]: it should be noted that these are just query samples and they can be specified in other ways too.

File: app-test-demo/README.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+* Feign - OpenFeign + OkHttp based client. Makes calls to backend API.
+* Assert - provides helper services to create data assertions and has visual representation in reports. 
+* Allure - report generation module. Based on Allure.
```
Comment (ricardas-buc) [inline app-test-demo/README.md]: If we provide link for testcontainers, we can do the same for other libraries

File: app-test-demo/src/test/java/com/nortal/test/demo/glue/TodoApiStepDefinitions.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+@RequiredArgsConstructor
+public class TodoApiStepDefinitions {
+    private static final Long TODO_ID = 1L;
+    private final TodoApiService todoApiService;
+    private final ScenarioContext scenarioContext;
+    private final DatabaseService databaseService;
+    private final TodoProvider todoProvider;
+    private final ValidationService validationService;
+
+    @Given("user has todos")
```
Comment (ricardas-buc) [inline app-test-demo/src/test/java/com/nortal/test/demo/glue/TodoApiStepDefinitions.java]: Framework provides a more generic annotation `@Step`. We should use that as there is no benefits of differentiating step types within definitions

File: app-test-demo/src/test/java/com/nortal/test/demo/glue/TodoApiStepDefinitions.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    public void setupTodosInDatabase() {
+        databaseService.addTodoToDatabase();
+    }
+
+    @When("user requests todo list")
+    public void getTodoList() {
+        ResponseEntity<List<Todo>> responseEntity = todoApiService.getTodoList();
```
Comment (ricardas-buc) [inline app-test-demo/src/test/java/com/nortal/test/demo/glue/TodoApiStepDefinitions.java]: In most cases there is no need to use step-data. See: https://github.com/nordic-institute/X-Road/tree/develop-7.x/src/centralserver/admin-service/int-test for better exaplanation

File: app-test-demo/src/test/java/com/nortal/test/demo/glue/TodoApiStepDefinitions.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+                .title("Validate delete todo response")
+                .assertion(statusCodeAssertion(HttpStatus.OK))
+                .assertion(new Assertion.Builder()
+                        .message("Response is empty")
+                        .expression("body")
+                        .operation(AssertionOperation.NULL).build());
+        validationService.validate(validationBuilder.build());
+    }
+
+    private Assertion equalsAssertion(Object expectedValue, String expression, String message) {
```
Comment (ricardas-buc) [inline app-test-demo/src/test/java/com/nortal/test/demo/glue/TodoApiStepDefinitions.java]: usually having parent class for all step definitions classes with helper methods is best way to handle these

File: app-test-demo/src/test/java/com/nortal/test/demo/hook/FrontendContainerSetup.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+import com.nortal.test.testcontainers.AbstractTestableContainerSetup;
+import com.nortal.test.testcontainers.images.builder.ImageFromDockerfile;
+import org.jetbrains.annotations.NotNull;
+import org.springframework.stereotype.Component;
+import org.testcontainers.images.builder.dockerfile.DockerfileBuilder;
+
+import java.util.Collections;
+import java.util.List;
+
+//@Component
```
Comment (ricardas-buc) [inline app-test-demo/src/test/java/com/nortal/test/demo/hook/FrontendContainerSetup.java]: is this still relevant?

**Review by ricardas-buc — CHANGES_REQUESTED** [review]

File: app-test-demo/src/test/java/com/nortal/test/demo/glue/TodoUiStepDefinitions.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    public static final String SAVE_TODO_BUTTON = "//div/form/button";
+    public static final String DESCRIPTION_VALUE = "test";
+    private final UiMediator uiMediator;
+    private final FrontendContextualContainer frontendContainer;
+    private int initialTodosCount;
+
+    @Step("user has opened Todo application in browser")
+    public void openPageInBrowser() {
+        Selenide.open(frontendContainer.getUiUrl());
+        uiMediator.elementXpath(MAIN_PAGE_HEADER, true).shouldBe(Condition.visible);
```
Comment (ricardas-buc) [inline app-test-demo/src/test/java/com/nortal/test/demo/glue/TodoUiStepDefinitions.java]: Use page object pattern per this example: 
https://github.com/nordic-institute/X-Road/tree/develop-7.x/src/centralserver/admin-service/ui-system-test/src/intTest/java/org/niis/xroad/cs/test/ui

We should also consider dropping ui mediator as it reduces  code readability

File: app-test-demo/src/test/resources/behavior/api/todo-api.feature [hunk 9]
```diff
@@ -0,0 +1,25 @@
+Feature: Demo Test - Todo API
+
+  """ This test is for demonstration purposes.
```
Comment (ricardas-buc) [inline app-test-demo/src/test/resources/behavior/api/todo-api.feature]: feature desc does not need quotes

**Review by ricardas-buc — CHANGES_REQUESTED** [review]

File: app-test-demo/src/test/resources/behavior/api/todo-api.feature (hunk 9 above)
Comment (vilmantas-r) [reply app-test-demo/src/test/resources/behavior/api/todo-api.feature]: fixed

File: app-test-demo/src/test/java/com/nortal/test/demo/glue/TodoUiStepDefinitions.java (hunk 8 above)
Comment (vilmantas-r) [reply app-test-demo/src/test/java/com/nortal/test/demo/glue/TodoUiStepDefinitions.java]: created PageObj classes, dropped ui mediator.

**Review by ricardas-buc — APPROVED** [review]

**Pull request merged**